gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1266 @@
+          "The 'parameters.managedZone' resource named '%s' does not exist.", zoneName));
+    }
+    ImmutableSortedMap<String, ResourceRecordSet> dnsRecords = zoneContainer.dnsRecords().get();
+    String[] fields = (String[]) options.get("fields");
+    String name = (String) options.get("name");
+    String type = (String) options.get("type");
+    String pageToken = (String) options.get("pageToken");
+    ImmutableSortedMap<String, ResourceRecordSet> fragment =
+        pageToken != null ? dnsRecords.tailMap(pageToken, false) : dnsRecords;
+    Integer maxResults = options.get("maxResults") == null
+        ? null : Integer.valueOf((String) options.get("maxResults"));
+    boolean sizeReached = false; // maximum result size was reached, we should not return more
+    boolean hasMorePages = false; // should next page token be included in the response?
+    LinkedList<String> serializedRrsets = new LinkedList<>();
+    String lastRecordId = null;
+    for (String recordId : fragment.keySet()) {
+      ResourceRecordSet record = fragment.get(recordId);
+      if (matchesCriteria(record, name, type)) {
+        if (sizeReached) {
+          // we do not add this, just note that there would be more and there should be a token
+          hasMorePages = true;
+          break;
+        } else {
+          lastRecordId = recordId;
+          try {
+            serializedRrsets.addLast(jsonFactory.toString(
+                OptionParsers.extractFields(record, fields)));
+          } catch (IOException e) {
+            return Error.INTERNAL_ERROR.response(String.format(
+                "Error when serializing resource record set in managed zone %s in project %s",
+                zoneName, projectId));
+          }
+        }
+      }
+      sizeReached = maxResults != null && maxResults.equals(serializedRrsets.size());
+    }
+    boolean includePageToken =
+        hasMorePages && (fields == null || ImmutableList.copyOf(fields).contains("nextPageToken"));
+    return toListResponse(serializedRrsets, "rrsets", lastRecordId, includePageToken);
+  }
+
+  /**
+   * Lists changes. Next page token is ID of the last change listed.
+   */
+  @VisibleForTesting
+  Response listChanges(String projectId, String zoneName, String query) {
+    Map<String, Object> options = OptionParsers.parseListChangesOptions(query);
+    Response response = checkListOptions(options);
+    if (response != null) {
+      return response;
+    }
+    ZoneContainer zoneContainer = findZone(projectId, zoneName);
+    if (zoneContainer == null) {
+      return Error.NOT_FOUND.response(String.format(
+          "The 'parameters.managedZone' resource named '%s' does not exist", zoneName));
+    }
+    // take a sorted snapshot of the current change list
+    NavigableMap<Integer, Change> changes = new TreeMap<>();
+    for (Change c : zoneContainer.changes()) {
+      if (c.getId() != null) {
+        changes.put(Integer.valueOf(c.getId()), c);
```

Comment:
I believe that this can throw a NumberFormatException because we add changes first before assigning them an ID.